embassy-stm32/src/timer/mod.rs

```diff
@@ -69,31 +69,40 @@ pub trait GeneralInstance1Channel: CoreInstance {}
 /// General-purpose 16-bit timer with 2 channels instance.
 pub trait GeneralInstance2Channel: GeneralInstance1Channel {}
 
-/// General-purpose 16-bit timer with 4 channels instance.
-pub trait GeneralInstance4Channel: BasicInstance + GeneralInstance2Channel {
+// This trait add *extra* methods to GeneralInstance4Channel,
+// that GeneralInstance4Channel doesn't use, but the "AdvancedInstance"s need.
+// And it's a private trait, so it's content won't leak to outer namespace.
+//
+// If you want to add a new method to it, please leave a detail comment to explain it.
+trait General4ChBlankSealed {
     // SimplePwm<'d, T> is implemented for T: GeneralInstance4Channel
     // Advanced timers implement this trait, but the output needs to be
     // enabled explicitly.
     // To support general-purpose and advanced timers, this function is added
     // here defaulting to noop and overwritten for advanced timers.
-    /// Enable timer outputs.
+    //
+    // Enable timer outputs.
     fn enable_outputs(&self) {}
 }
 
+/// General-purpose 16-bit timer with 4 channels instance.
+#[allow(private_bounds)]
+pub trait GeneralInstance4Channel: BasicInstance + GeneralInstance2Channel + General4ChBlankSealed {}
+
 /// General-purpose 32-bit timer with 4 channels instance.
 pub trait GeneralInstance32bit4Channel: GeneralInstance4Channel {}
 
 /// Advanced 16-bit timer with 1 channel instance.
-pub trait AdvancedInstance1Channel: BasicNoCr2Instance + GeneralInstance1Channel {
-    /// Capture compare interrupt for this timer.
-    type CaptureCompareInterrupt: interrupt::typelevel::Interrupt;
-}
+pub trait AdvancedInstance1Channel: BasicNoCr2Instance + GeneralInstance1Channel {}
 /// Advanced 16-bit timer with 2 channels instance.
 
 pub trait AdvancedInstance2Channel: BasicInstance + GeneralInstance2Channel + AdvancedInstance1Channel {}
 
 /// Advanced 16-bit timer with 4 channels instance.
-pub trait AdvancedInstance4Channel: AdvancedInstance2Channel + GeneralInstance4Channel {}
+pub trait AdvancedInstance4Channel: AdvancedInstance2Channel + GeneralInstance4Channel {
+    /// Capture compare interrupt for this timer.
+    type CaptureCompareInterrupt: interrupt::typelevel::Interrupt;
```

Comment:
why move the interrupt from the 1ch to the 4ch trait? 1ch/2ch timers still have the interrupt, right?

(not that it matters now, because we're implementing all 3 traits for all timers ignoring the channel count, but it'll matter after we fix that)

Reply:
but they still "logically" have one, it's just that on some timers the CC and UP interrupts go to the same NVIC line? 

it's the same as the I2C EV and ER interrupts. On some chips they go to a single NVIC interrupt I2C1, and in others it's separated into I2C1_EV, I2C1_ER. On the Embassy side we represent it as two `type xxInterrupt` associated types, which will be the same interrupt on chips that don't split it. This works on all chips, since `bind_interrupts!` allows binding the same interrupt to multiple handlers.

though then perhaps with that logic the CC interrupt should go in the "general" instance traits since these can also do CC, hmm.

Reply:
When it first introduced in #2614, CC is attached to AdvTim (now Adv4Ch).

So for current situation, we should move it to GeneralInstance1Channel?